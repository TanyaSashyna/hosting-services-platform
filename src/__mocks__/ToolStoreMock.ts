/* eslint-disable @typescript-eslint/no-explicit-any  */
/* eslint-disable @typescript-eslint/no-empty-function  */

import { ToolStore } from "../stores/ToolStore";
import { StatusTool, DnsChangesType } from "../api/ToolApi";

class ToolStoreMock extends ToolStore {
  getToolStatus(): any {}
  addTool(): any {}
  cancelTool(): any {}
  status: StatusTool = "disabled";
  domain = "";
  dnsChangesRequired = true;
  dnsChanges: DnsChangesType[] = [];
  percentage = 0;
  hostReachable = true;
}

const toolStoreMock = new ToolStoreMock();

toolStoreMock.getToolStatus = jest.fn();
toolStoreMock.addTool = jest.fn();
toolStoreMock.cancelTool = jest.fn();

export default toolStoreMock;
