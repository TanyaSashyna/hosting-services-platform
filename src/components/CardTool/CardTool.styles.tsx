import styled from "styled-components";
import {
  Card as _Card,
  breakpoints,
  fontWeight,
} from "@project-name-frontend/ui-package";

export const Card = styled(_Card)`
  .cardInfo:first-child {
    @media (max-width: ${breakpoints.phoneUpperBoundary}) {
      padding-bottom: 0;
    }
  }

  .card-align {
    align-items: center;
  }

  .thin-text {
    font-weight: ${fontWeight.regular};
  }
`;
