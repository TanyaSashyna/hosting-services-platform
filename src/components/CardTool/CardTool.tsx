import { inject, observer } from "mobx-react";
import React, { FC } from "react";
import {
  CardRow,
  CardInfo,
  useMobile,
  Button,
  Tooltip,
} from "@project-name-frontend/ui-package";

import { Card } from "./CardTool.styles";
import { CardNotificationBanner } from "../CardNotificationBanner";
import { ToolStore } from "../../stores/toolStore";

type InjectedProps = {
  toolStore: ToolStore;
};

type CardToolProps = {
  domain: string;
  role?: string;
  isError?: boolean;
  loading: boolean;
  toolClearCacheEnabled: boolean;
  setIsOpenAddToolModal: (state: boolean) => void;
  setIsCancelToolModal: (state: boolean) => void;
};

export const CardTool: FC<CardToolProps & InjectedProps> = ({
  toolStore,
  domain,
  role,
  isError,
  loading,
  toolClearCacheEnabled,
  setIsOpenAddToolModal,
  setIsCancelToolModal,
}) => {
  const { mobile } = useMobile();
  const { status, hostReachable, canChangeDns } = toolStore;
  const enabled = status === "enabled";
  const inProgressStatus = status === "inProgress" || status === "created";

  return (
    <>
      <Card
        header={
          enabled ? (
            `Cancel Tool for ${domain}`
          ) : (
            <div>
              {`Add Tool to ${domain}`}
              <div>29 per month</div>
            </div>
          )
        }
        noPad={true}
      >
        <CardRow
          direction={mobile ? "column" : "row"}
          spacing="space-between"
          className={!mobile ? "card-align" : ""}
        >
          <CardInfo
            content={{
              children: enabled
                ? "Tool optimises website performance and speed."
                : "Speed up content delivery on your website.",
            }}
          />
          <CardInfo
            align={mobile ? "center" : "right"}
            action={{
              fullWidth: true,
              children: enabled ? (
                <Tooltip
                  placement="top"
                  title={
                    role === "TechnAdmin"
                      ? "You don't have the ability to cancel Tool"
                      : ""
                  }
                >
                  <Button
                    fullWidth={mobile}
                    buttonType="secondary"
                    onClick={(): void => setIsCancelToolModal(true)}
                    data-testid="cancel-btn"
                    disabled={loading || role === "TechnAdmin"}
                  >
                    Cancel service
                  </Button>
                </Tooltip>
              ) : (
                <Tooltip
                  placement="top"
                  title={
                    !isError && inProgressStatus
                      ? "Setting up Tool"
                      : role === "TechnAdmin"
                      ? "You don't have the ability to activate Tool"
                      : ""
                  }
                >
                  <Button
                    fullWidth={mobile}
                    leftIcon="add"
                    disabled={
                      !hostReachable ||
                      (!isError && inProgressStatus) ||
                      role === "TechnAdmin"
                    }
                    onClick={(): void => setIsOpenAddToolModal(true)}
                    data-testid="add-btn"
                  >
                    Add
                  </Button>
                </Tooltip>
              ),
            }}
          />
        </CardRow>
        <CardNotificationBanner
          hostReachable={hostReachable}
          status={status}
          canChangeDns={canChangeDns}
        />
      </Card>
      {enabled && !toolClearCacheEnabled && (
        <Card
          header="Edge cache"
          subtitle="Determines how long content stays in the cache."
          noPad={true}
        >
          <CardRow>
            <CardInfo
              header={{
                children: (
                  <>
                    1 hour
                    <span className="thin-text">{" (default)"}</span>
                  </>
                ),
              }}
            />
          </CardRow>
        </Card>
      )}
    </>
  );
};

CardTool.displayName = "CardTool";

export default inject("toolStore")(
  observer(CardTool)
) as unknown as FC<CardToolProps>;
