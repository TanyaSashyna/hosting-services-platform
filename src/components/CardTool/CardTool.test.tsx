import React from "react";
import { cleanup, render, fireEvent, waitFor } from "@testing-library/react";

import { CardTool } from "./CardTool";
import ToolStoreMock from "../../__mocks__/ToolStoreMock";

afterEach(cleanup);

describe("CardTool", () => {
  const domainMock = "example.com";
  const mockFunc = jest.fn();

  it("renders CardTool and matches snapshot", () => {
    const { container, asFragment } = render(
      <CardTool
      setIsOpenAddToolModal={mockFunc}
      setIsCancelToolModal={mockFunc}
        domain={domainMock}
        toolStore={ToolStoreMock}
        loading={false}
        toolClearCacheEnabled={true}
      />
    );

    expect(container).toBeDefined();
    expect(asFragment()).toMatchSnapshot();
  });

  it("trigger setIsOpenAddToolModal function on Add button click", () => {
    const { getByText, getByTestId } = render(
      <CardTool
        setIsOpenAddToolModal={mockFunc}
        setIsCancelToolModal={mockFunc}
        domain={domainMock}
        toolStore={ToolStoreMock}
        loading={false}
        toolClearCacheEnabled={true}
      />
    );

    const addBtn = getByTestId("add-btn");
    fireEvent.click(addBtn);

    expect(getByText("29 per month")).toBeDefined();
    expect(getByText("Add Tool to example.com")).toBeDefined();
    expect(
      getByText("Speed up content delivery on your website.")
    ).toBeDefined();
    expect(mockFunc).toHaveBeenCalled();
  });

  it("trigger setIsCancelToolModal function on Cancel button click", () => {
    ToolStoreMock.status = "enabled";

    const { getByText, getByTestId } = render(
      <CardTool
        setIsOpenAddToolModal={mockFunc}
        setIsCancelToolModal={mockFunc}
        domain={domainMock}
        toolStore={ToolStoreMock}
        loading={false}
        toolClearCacheEnabled={true}
      />
    );

    const cancelBtn = getByTestId("cancel-btn");
    fireEvent.click(cancelBtn);

    expect(getByText("Cancel Tool for example.com")).toBeDefined();
    expect(mockFunc).toHaveBeenCalled();
  });

  it("renders when toolClearCacheEnabled is false", () => {
    ToolStoreMock.status = "enabled";

    const { getByText } = render(
      <CardTool
        setIsOpenAddToolModal={mockFunc}
        setIsCancelToolModal={mockFunc}
        domain={domainMock}
        toolStore={ToolStoreMock}
        loading={false}
        toolClearCacheEnabled={false}
      />
    );

    expect(getByText("Edge cache")).toBeDefined();
  });

  it("renders when Cancel service button is disabled", () => {
    ToolStoreMock.status = "enabled";

    const { getByTestId } = render(
      <CardTool
        setIsOpenAddToolModal={mockFunc}
        setIsCancelToolModal={mockFunc}
        domain={domainMock}
        toolStore={ToolStoreMock}
        loading={true}
        toolClearCacheEnabled={true}
      />
    );

    const cancelBtn = getByTestId("cancel-btn");
    expect(cancelBtn).toHaveProperty("disabled", true);
  });

  it("renders when setting up Tool", () => {
    ToolStoreMock.status = "inProgress";

    const { getByTestId, getByText } = render(
      <CardTool
        setIsOpenAddToolModal={mockFunc}
        setIsCancelToolModal={mockFunc}
        domain={domainMock}
        toolStore={ToolStoreMock}
        loading={false}
        toolClearCacheEnabled={true}
      />
    );

    const addBtn = getByTestId("add-btn");
    fireEvent.mouseOver(addBtn);

    expect(waitFor(() => expect(getByText("Setting up Tool")))).toBeDefined();
  });

  it("renders when the role is a TechnAdmin and the status is disabled", () => {
    ToolStoreMock.status = "disabled";

    const { getByTestId, getByText } = render(
      <CardTool
        setIsOpenAddToolModal={mockFunc}
        setIsCancelToolModal={mockFunc}
        domain={domainMock}
        toolStore={ToolStoreMock}
        loading={false}
        toolClearCacheEnabled={true}
        role="TechnAdmin"
      />
    );

    const addBtn = getByTestId("add-btn");
    fireEvent.mouseOver(addBtn);

    expect(
      waitFor(() =>
        expect(getByText("You don't have the ability to activate Tool"))
      )
    ).toBeDefined();
  });

  it("renders when the role is a TechnAdmin and the status is enabled", () => {
    ToolStoreMock.status = "enabled";

    const { getByTestId, getByText } = render(
      <CardTool
        setIsOpenAddToolModal={mockFunc}
        setIsCancelToolModal={mockFunc}
        domain={domainMock}
        toolStore={ToolStoreMock}
        loading={false}
        toolClearCacheEnabled={true}
        role="TechnAdmin"
      />
    );

    const cancelBtn = getByTestId("cancel-btn");
    fireEvent.mouseOver(cancelBtn);

    expect(
      waitFor(() =>
        expect(getByText("You don't have the ability to cancel Tool"))
      )
    ).toBeDefined();
  });
});
