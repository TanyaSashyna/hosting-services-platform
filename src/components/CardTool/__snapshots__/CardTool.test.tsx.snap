// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardTool renders CardTool and matches snapshot 1`] = `
<DocumentFragment>
  .c0 .card-align {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 .thin-text {
  font-weight: 400;
}

.c1 {
  padding: 16px;
  padding-top: 0;
}

@media (max-width:768px) {
  .c0 .cardInfo:first-child {
    padding-bottom: 0;
  }
}

<div
    class="card c0"
  >
    <div
      class="card__head"
    >
      <div
        class="card__head__text"
      >
        <div
          class="card__head--wrap"
        >
          <div
            class="card__title"
          >
            <div>
              Add Tool to example.com
              <div>
                R29 per month.
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="card__body card__body--no-pad"
    >
      <div
        class="cardRow cardRow--direction-column cardRow--spacing-space-between"
      >
        <div
          class="cardInfo"
        >
          <div
            class="cardInfo__inner"
          >
            <span
              class="cardInfo__content"
            >
              Speed up content delivery on your website.
            </span>
          </div>
        </div>
        <div
          class="cardInfo cardInfo--align-center"
        >
          <div
            class="cardInfo__inner cardInfo--fullWidth"
          >
            <div
              class="cardInfo__action"
            >
              <button
                class="button button--size-default button--type-default button--text button--fullWidth"
                data-testid="add-btn"
                title=""
                type="button"
              >
                <div
                  class="button__internal"
                >
                  <div
                    class="button__icon"
                  >
                    <i
                      class="icon"
                    >
                      add
                    </i>
                  </div>
                  <div
                    class="button__children"
                  >
                    Add
                  </div>
                </div>
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="c1"
      >
        <div
          class="notificationBanner notificationBanner__border--neutral notificationBanner--borderAll"
        >
          <div
            class="notificationBanner__icon"
          >
            <i
              class="icon icon--neutral"
            >
              info
            </i>
          </div>
          <div
            class="notificationBanner__content"
          >
            <div
              class="notificationBanner__content__title"
            >
              Note
            </div>
            <div
              class="notificationBanner__content__body"
            >
              Adding Tool to your domain automatically updates your DNS records to point to Tool.
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
